features/steps/steps.py

```diff
@@ -200,6 +209,61 @@ def step_impl(context, constraint, num, entity):
 
     handle_errors(context, errors)
 
+@then('Each {entity} must nest {constraint} {num:d} instance(s) of {other_entity}')
+def step_impl(context, entity, num, constraint, other_entity):
+    stmt_to_op = {'exactly': operator.eq, "at most": operator.le}
+    assert constraint in stmt_to_op
+    op = stmt_to_op[constraint]
+
+    errors = []
+
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            nested_entities = [entity for rel in inst.IsNestedBy for entity in rel.RelatedObjects]
+            if not op(sum(1 for i in nested_entities if i.is_a() == other_entity), num):
+                errors.append(instance_count_error([i for i in nested_entities if i.is_a() == other_entity]))
+
+    handle_errors(context, errors)
+
+@then('Each {entity} must be nested only by {other_entity}')
+def step_impl(context, entity, other_entity):
+    
+    errors = []
+
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            relating_objects_lst = [i.RelatingObject.is_a() for i in inst.Nests]
+            relating_object = relating_objects_lst[0]
+            if not other_entity == relating_object:
+                errors.append(instance_nesting_error(inst, relating_objects_lst))
+
+@then('Each {entity} must nest only the following entities: {other_entities}')
+def step_impl(context, entity, other_entities):
+
+    errors = []
+
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            nested_entities = [i for rel in inst.IsNestedBy for i in rel.RelatedObjects]
+            if not set([i.is_a() for i in nested_entities]).issubset(other_entities.split(',')):
```

Comment:
As a team we need to decide on using

- `instance.is_a(str)`, or
- `instance.is_a() == str`, or
- `instance.is_a().lower() == str.lower()`

The first takes into account (a) inheritance and (b) difference in case and (c) will throw an error when `str` is not a valid entity name given the file schema of `instance`. The second takes into account nothing. The third takes into account differences in case.

Reply:
A comment on `issubset()` (that I had to look up) it actually also returns true when the sets are equal. For that reason I prefer writing `<=` (subset or equal) because it's a better differentiated from `<` (strict subset).

Reply:
You're right, it should return True if the two sets are equal. A better name (logically) would then be .isorequaltosubset but I can see why they did not pick such a name and went for .is instead .. <= is short and clear

instance.is_a(str) seems to be the most comprehensive you how describe it. Maybe I am understanding it wrong, but in case of (c) it just returns False (e.g.  ifc_file.by_type("IfcAlignment")[0].is_a('IfcNotExistingBlob') returns False) but not an error?